lib/csv.rb

```diff
@@ -1592,14 +1604,23 @@ def open(filename, mode="r", **options)
       options.delete(:replace)
       options.delete_if {|k, _| /newline\z/.match?(k)}
 
-      begin
-        f = File.open(filename, mode, **file_opts)
-      rescue ArgumentError => e
-        raise unless /needs binmode/.match?(e.message) and mode == "r"
-        mode = "rb"
-        file_opts = {encoding: Encoding.default_external}.merge(file_opts)
-        retry
+      if filename_or_io.is_a?(StringIO)
+        f = if RUBY_VERSION >= "2.7"
+          StringIO.new(filename_or_io.string, **file_opts)
+        else
+          StringIO.new(filename_or_io.string)
+        end
```

Comment:
Can we always use an empty `file_opts` for Ruby 2.7 or earlier instead?

```diff
diff --git a/lib/csv.rb b/lib/csv.rb
index 5f75894..6f94bd4 100644
--- a/lib/csv.rb
+++ b/lib/csv.rb
@@ -1605,11 +1605,7 @@ class CSV
       options.delete_if {|k, _| /newline\z/.match?(k)}
 
       if filename_or_io.is_a?(StringIO)
-        f = if RUBY_VERSION >= "2.7"
-          StringIO.new(filename_or_io.string, **file_opts)
-        else
-          StringIO.new(filename_or_io.string)
-        end
+        f = StringIO.new(filename_or_io.string, **file_opts)
       else
         begin
           f = File.open(filename_or_io, mode, **file_opts)
@@ -1911,7 +1907,11 @@ class CSV
                                                mode,
                                                options,
                                                file_opts)
-      unless filename_or_io.is_a?(StringIO)
+      if filename_or_io.is_a?(StringIO)
+        # Support to StringIO was dropped for Ruby 2.6 and earlier without BOM support:
+        # https://github.com/ruby/stringio/pull/47
+        return if RUBY_VERSION < "3.0"
+      else
         # "bom|utf-8" may be buggy on Windows:
         # https://bugs.ruby-lang.org/issues/20526
         return if ON_WINDOWS
```

Reply:
This doesn't work because `CSV#open` still adds keys to `file_opts` here: https://github.com/ruby/csv/blob/bb93c289a529c77b5d5a8788dd8fdc03052216fc/lib/csv.rb#L1586-L1589

For Ruby < 2.7, if `file_opts` is not empty, we get `TypeError: no implicit conversion of Hash into String`.

I like that something fails if `options` is given to `CSV#open` when it's not supported, as opposed to simply ignoring it, the way I did it, but then the error should be ideally more clear than what the old `StringIO` returns. I can go that route, by rescuing the `TypeError`, and raising another exception with a clearer message. In any case, the `file_opts[:universal_newline] ||= false` would still have to be removed for `< 2.7`.

Reply:
> This doesn't work because `CSV#open` still adds keys to `file_opts` here:
> 
> https://github.com/ruby/csv/blob/bb93c289a529c77b5d5a8788dd8fdc03052216fc/lib/csv.rb#L1586-L1589

Oh, sorry.

> For Ruby < 2.7, if `file_opts` is not empty, we get `TypeError: no implicit conversion of Hash into String`.
> 
> I like that something fails if `options` is given to `CSV#open` when it's not supported, as opposed to simply ignoring it, the way I did it, but then the error should be ideally more clear than what the old `StringIO` returns. I can go that route, by rescuing the `TypeError`, and raising another exception with a clearer message.

I agree with you but can we use `raise ... unless file_opts.empty?` for it instead of rescuing the `TypeError`?

> In any case, the `file_opts[:universal_newline] ||= false` would still have to be removed for `< 2.7`.

I agree with you.

Reply:
It turns out passing an empty `file_opts` doesn't work either. I've noticed this quirk from the interpreter in Ruby 2.6:

```ruby
StringIO.new("bla", **{}) # => #<StringIO:0x0000563108bb88b8>

opts = {}
StringIO.new("bla", **opts) # => TypeError (no implicit conversion of Hash into String)
```

To address that, I've changed the `StringIO` initialization in Ruby < 2.7, so that it:
- ignores options that are meant only for `CSV.new`
- raises an `ArgumentError` if anything else is passed as `options` to `CSV.open`
- initializes `StringIO` without keyword arguments.

Reply:
OK. Let's use the approach.